external/swloop/swloop.cpp

```diff
@@ -0,0 +1,533 @@
+                    return;
+                }
+                Eigen::MatrixXcd hamtol = ham + (Eigen::MatrixXcd::Identity(nHam, nHam) * tol0);
+                chol.compute(hamtol);
+                if (chol.info() == Eigen::NumericalIssue) {
+                    chol.compute(ham + (Eigen::MatrixXcd::Identity(nHam, nHam) * params.omega_tol));
+                    if (chol.info() == Eigen::NumericalIssue) {
+                        err_flag.store(2, std::memory_order_relaxed);
+                        return;
+                    }
+                }
+                outputs.warn_posdef = true;
+            }
+            Eigen::MatrixXcd K = chol.matrixU();
+            Eigen::MatrixXcd K2 = K * gComm * K.adjoint().eval();
+            K2 = (K2 + K2.adjoint().eval()) / 2;
+            Eigen::SelfAdjointEigenSolver<Eigen::MatrixXcd> eig(K2);
+            if (eig.info() != Eigen::Success) {
+                err_flag.store(3, std::memory_order_relaxed);
+                return;
+            }
+            std::vector<ev_tuple_t> evs;
+            for (size_t ii=0; ii<nHam; ii++) {
+                evs.push_back(ev_tuple_t(eig.eigenvalues()[ii], eig.eigenvectors().col(ii))); }
+            std::sort(evs.begin(), evs.end(),
+                [&](const ev_tuple_t &a, const ev_tuple_t &b) -> bool {
+                    return std::real(std::get<0>(a)) > std::real(std::get<0>(b)); });
+            Eigen::MatrixXcd U(nHam, nHam);
+            for (size_t ii=0; ii<nHam; ii++) {
+                *(omega++) = std::get<0>(evs[ii]);  // Note this pointer arithmetic assumes column-major ordering
+                U.col(ii) = std::get<1>(evs[ii]) * sqrt(gComm(ii,ii) * std::get<0>(evs[ii]));
+            }
+            V = K.inverse() * U;
+        } else {
+            Eigen::MatrixXcd gham = (gComm * ham).eval();
+            // Add a small amount to the diagonal to ensure there are no degenerate levels,
+            // so that the eigenvectors would be (quasi)orthogonal (instead of using eigorth()).
+            double vd = 0.5 - (double)params.nMagExt;
+            for (size_t ii=0; ii<nHam; ii++) { gham(ii, ii) += vd*1e-12; vd++; }
+            Eigen::ComplexEigenSolver<Eigen::MatrixXcd> eig(gham);
+            if (eig.info() != Eigen::Success) {
+                err_flag.store(3, std::memory_order_relaxed);
+                return;
+            }
+            std::vector<ev_tuple_t> evs;
+            for (size_t ii=0; ii<nHam; ii++) {
+                evs.push_back(ev_tuple_t(eig.eigenvalues()[ii], eig.eigenvectors().col(ii))); }
+            std::sort(evs.begin(), evs.end(),
+                [&](const ev_tuple_t &a, const ev_tuple_t &b) -> bool {
+                    return std::real(std::get<0>(a)) > std::real(std::get<0>(b)); });
+            Eigen::MatrixXcd U(nHam, nHam);
+            for (size_t ii=0; ii<nHam; ii++) {
+                *(omega++) = std::get<0>(evs[ii]);  // Note this pointer arithmetic assumes column-major ordering
```

Comment:
This pointer arithmetic looks very clever but difficult to interpret... is it assigning energy eigenvalues? Maybe a comment to explain?

Reply:
Yes, it's assigning every energy eigenvalue to the array pointed to by `omega`. At the start `omega` points to the first element of the array. `*omega` dereferences the pointer and assigns the first element. `omega++` increments the pointer to point to the next element after the assignment. Because the `omega` array is actually 2D, where each column represents the magnon energies of each q-point, and we are running the calculation q-point by q-point, this assignment only works if the memory layout of `omega` is column-major (although for row-major layouts we can do a transpose at the end I guess).